power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,114 +25,242 @@ namespace detail = power_grid_model::optimizer::detail;
+
+    auto operator<=>(const TrafoGraphEdge& other) const {
+        if (auto cmp = weight <=> other.weight; cmp != 0) { // NOLINT(modernize-use-nullptr)
+            return cmp;
+        }
+        if (auto cmp = regulated_idx.group <=> other.regulated_idx.group; cmp != 0) { // NOLINT(modernize-use-nullptr)
+            return cmp;
+        }
+        return regulated_idx.pos <=> other.regulated_idx.pos;
+    }
+};
+
+using TrafoGraphEdges = std::vector<std::pair<TrafoGraphIdx, TrafoGraphIdx>>;
+using TrafoGraphEdgeProperties = std::vector<TrafoGraphEdge>;
+using RankedTransformerGroups = std::vector<Idx2D>;
+
+struct RegulatedObjects {
+    std::set<Idx> transformers{};
+    std::set<Idx> transformers3w{};
 };
 
-// TODO(mgovers): investigate whether this really is the correct graph structure
 using TransformerGraph = boost::compressed_sparse_row_graph<boost::directedS, TrafoGraphVertex, TrafoGraphEdge,
                                                             boost::no_property, TrafoGraphIdx, TrafoGraphIdx>;
 
+template <std::derived_from<ThreeWindingTransformer> Component, class ComponentContainer>
+    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
+constexpr void add_edge(main_core::MainModelState<ComponentContainer> const& state,
+                        RegulatedObjects const& regulated_objects, TrafoGraphEdges& edges,
+                        TrafoGraphEdgeProperties& edge_props) {
+    // Only add idx2d to one of the 2 directional edges of 3 winding transformer which ensures a single ranking
+    bool assigned_idx{};
+    using enum Branch3Side;
+    std::array<std::tuple<Branch3Side, Branch3Side>, 3> const branch3_combinations{
+        {{side_1, side_2}, {side_2, side_3}, {side_1, side_3}}};
+    for (auto const& transformer3w : state.components.template citer<ThreeWindingTransformer>()) {
+        for (auto const& [first_side, second_side] : branch3_combinations) {
+            if (!transformer3w.status(first_side) || !transformer3w.status(second_side)) {
+                continue;
+            }
+            auto const& from_node = transformer3w.node(first_side);
+            auto const& to_node = transformer3w.node(second_side);
+
+            auto const tap_at_first_side = transformer3w.tap_side() == first_side;
+            auto const single_direction_condition = regulated_objects.transformers3w.contains(transformer3w.id()) &&
+                                                    (tap_at_first_side || transformer3w.tap_side() == second_side);
+            if (single_direction_condition) {
+                auto const& tap_side_node = tap_at_first_side ? from_node : to_node;
+                auto const& non_tap_side_node = tap_at_first_side ? to_node : from_node;
+                Idx2D const regulated_idx =
+                    assigned_idx ? unregulated_idx : main_core::get_component_idx_by_id(state, transformer3w.id());
+                assigned_idx = true;
```

Comment:
wouldn't this logic be much easier if you would loop over `{(side_1, side_2), (side_2, side_3), (side_3, side_1)}` instead of `(side_1, side_3)` for the last element?

you can then check whether `first_side` is regulated. if so, assign `regulated_idx`. if not, assign `unregulated_idx`. The fact that the loop contains cyclic combinations ensures uniqueness. no need for complex logic like `assigned_idx`